### ISSUE TITLE
tonlib-cuda-cli doens't always terminate itself when having error

### ISSUE THREAD

**Issue by winiel559** [tontechio/pow-miner-gpu#64]
Hello, I'm using the systemd automation method described in pow-miner-howto.md. Ubuntu 18.04 (HiveOS)
I'm running several mining instances at once.

One tonlib-cuda-cli instance got an error,
but it didn't terminate itself, so the systemd couldn't restart it.
I found the error 20min later and restarted it manually.
Here's the log:

> 00:07:53 [tonlib-cli.cpp:751][!PowMiner][&status.is_error()] pminer: [Error : 500 : LITE_SERVER_NETWORKadnl query timeout]
> 00:27:49 systemd[1]: Stopping NewTON miner0...

Sometime it works (the error message is slightly different though) 

> 00:52:11 tonlib-cuda-cli[10005]: [ 1][t 1][2021-11-16 16:52:11.355709890][tonlib-cli.cpp:751][!PowMiner][&stat.is_error()]        pminer: [Error : 500 : LITE_SERVER_NETWORKadnl query timeout(during last block synchronization)]
> 00:52:16 tonlib-cuda-cli[10005]: [pid 10005] [time 1637081536] Signal: 11
> 00:52:16 tonlib-cuda-cli[10005]: ------- Stack Backtrace -------
> 00:52:16 tonlib-cuda-cli[10005]: /usr/bin/ton/tonlib/tonlib-cuda-cli(+0x277a71)[0x55f055077a71]
> 00:52:16 tonlib-cuda-cli[10005]: /usr/bin/ton/tonlib/tonlib-cuda-cli(+0x277d81)[0x55f055077d81]
> 00:52:16 tonlib-cuda-cli[10005]: /usr/bin/ton/tonlib/tonlib-cuda-cli(+0x274e98)[0x55f055074e98]
> 00:52:16 tonlib-cuda-cli[10005]: /lib/x86_64-linux-gnu/libpthread.so.0(+0x12980)[0x7f15cf5c1980]
> 00:52:16 tonlib-cuda-cli[10005]: /lib/x86_64-linux-gnu/libpthread.so.0(pthread_rwlock_wrlock+0x12)[0x7f15cf5bb
> 00:52:16 tonlib-cuda-cli[10005]: /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1(CRYPTO_THREAD_write_lock+0x9)[0x7f
> 00:52:16 tonlib-cuda-cli[10005]: /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1(RAND_get_rand_method+0x33)[0x7f15c
> 00:52:16 tonlib-cuda-cli[10005]: /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1(RAND_bytes+0x10)[0x7f15cf96e1b0]
> 00:52:16 tonlib-cuda-cli[10005]: /usr/bin/ton/tonlib/tonlib-cuda-cli(+0x2a24b6)[0x55f0550a24b6]
> 00:52:16 tonlib-cuda-cli[10005]: /usr/bin/ton/tonlib/tonlib-cuda-cli(+0x315524)[0x55f055115524]
> 00:52:16 tonlib-cuda-cli[10005]: /usr/bin/ton/tonlib/tonlib-cuda-cli(+0x25397f)[0x55f05505397f]
> 00:52:16 tonlib-cuda-cli[10005]: /usr/bin/ton/tonlib/tonlib-cuda-cli(+0x253df0)[0x55f055053df0]
> 00:52:16 tonlib-cuda-cli[10005]: /usr/bin/ton/tonlib/tonlib-cuda-cli(+0x2675e5)[0x55f0550675e5]
> 00:52:16 tonlib-cuda-cli[10005]: /usr/bin/ton/tonlib/tonlib-cuda-cli(+0x2581a8)[0x55f0550581a8]
> 00:52:16 tonlib-cuda-cli[10005]: /usr/bin/ton/tonlib/tonlib-cuda-cli(main+0xe77)[0x55f054f286b7]
> 00:52:16 tonlib-cuda-cli[10005]: /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7)[0x7f15ce8a0bf7]
> 00:52:16 tonlib-cuda-cli[10005]: /usr/bin/ton/tonlib/tonlib-cuda-cli(_start+0x2a)[0x55f054f2dfda]
> 00:52:16 tonlib-cuda-cli[10005]: -------------------------------
> 00:52:16 systemd[1]: miner2.service: Main process exited, code=exited, status=1/FAILURE
> 00:52:16 systemd[1]: miner2.service: Failed with result 'exit-code'.
> 00:52:21 systemd[1]: miner2.service: Service hold-off time over, scheduling restart.
> 00:52:21 systemd[1]: miner2.service: Scheduled restart job, restart counter is at 4.
> 00:52:21 systemd[1]: Stopped NewTON miner2.

Here is one systemd service file if it helps
(I only change device id and name for different instance)
Thank you!

> [Unit]
> Description=NewTON miner0
> After=network.target
> 
> [Service]
> RestartSec=5
> Restart=always
> WorkingDirectory=/home/user
> ExecStartPre=/bin/sleep 20
> ExecStart=/usr/bin/ton/tonlib/tonlib-cuda-cli -v 3 -C /usr/bin/ton/tonlib/global.config.json  -e 'pminer start [giver addr] [my wallet] 3 16'
> 
> [Install]
> WantedBy=multi-user.target
> Alias=miner0.service

**Comment by optimizmo** [comment]
I had a similar issue, but under a different circumstances. I got
`[ 1][t 1][2021-11-17 11:13:11.484230695][tonlib-cli.cpp:755][!PowMiner] pminer: [Error : 500 : LITE_SERVER_UNKNOWN: timeout(during last block synchronization)] (#2276831547)`
when running under `supervisord` and the supervisor didn't try to restart the worker while it was just hanging. I killed the process manually and it then was restarted as expected by the supervisor.
